steward/src/cork/proposals.rs

```diff
@@ -0,0 +1,372 @@
+                }
+                let encoded_call: Vec<u8> = match governance_call.call.unwrap() {
+                    Call::AaveV2Stablecoin(data) => {
+                        if data.function.is_none() {
+                            error!(
+                                // if this happens there is a problem with proposal validation in sommelier
+                                "cork proposal {} call data contains no function data: {:?}",
+                                proposal.proposal_id, data,
+                            );
+                            continue;
+                        }
+                        let function = data.function.unwrap();
+                        aave_v2_stablecoin::get_encoded_governance_call(
+                            function,
+                            &cellar_id,
+                            proposal.proposal_id,
+                        )?
+                    }
+                    Call::CellarV1(data) => {
+                        if data.function.is_none() {
+                            error!(
+                                "cork proposal {} call data contains no function data: {:?}",
+                                proposal.proposal_id, data
+                            );
+                            continue;
+                        }
+                        let function = data.function.unwrap();
+                        cellar_v1::get_encoded_governance_call(
+                            function,
+                            &cellar_id,
+                            proposal.proposal_id,
+                        )?
+                    }
+                };
+                debug!(
+                    "proposal {} encoded call: {:?}",
+                    proposal.proposal_id, encoded_call
+                );
+
+                if let Err(err) =
+                    schedule_cork(&cellar_id, encoded_call.clone(), block_height).await
+                {
+                    error!(
+                        "failed to schedule cork for proposal {}, will retry shortly. reason: {}",
+                        proposal.proposal_id, err
+                    );
+                    // don't add an item twice to retry data
+                    for d in state.retry_data.iter() {
+                        if d.0 == proposal.proposal_id {
+                            continue;
+                        }
+                    }
+                    state.retry_data.push((
+                        proposal.proposal_id,
+                        cellar_id.clone(),
+                        encoded_call,
+                        block_height,
+                        0,
+                    ));
+                }
+                info!(
+                    "submitted cork for cellar {} triggered by proposal {}",
+                    cellar_id, proposal.proposal_id
+                );
+            }
+        }
+        pagination.offset += processed_total;
+    }
+
+    retry_failed_schedulings(
```

Comment:
Some thoughts here. The current implementation is retrieving every proposal with pagination, storing failed state about each proposal, and then handing these retries asynchronously. I think a simpler way to handle this whole process would just be this:

1. Add the latest observed proposal ID as part of the state.
2. Rather than querying all proposals, query individual proposal IDs in an incrementing loop. You'll have reached the end when it doesn't exist for your incremented ID.
3. In this method, you won't have to decouple retries -- just do retries inline before moving on to the next proposal ID.

Reply:
Yeah that's much better.